### PR TITLE
Mat 403. Add off heap memory manager.

### PR DESCRIPTION
**Pull request by stuartarchibald**
**This PR:**
- Adds the off heap aligned memory manager to OG-Maths.
- Adds windows port of the above (no `posix_memalign()` on windows).
- Is a stub more more general off heap memory operations.
- Gets ticket [[MAT-414]](http://jira.opengamma.com/browse/MAT-414) for addition work to gracefully handle exits. Rather suspect an API change will be required so a "status" can propagate to a C++ handler which then throws exceptions back to java.

**Coverage** (bearing in mind we are covering OS level stuff):
Java: 76% (up 3%, new code covered to 99%)
build/src/jshim 0.8% (no change)
build/src/librdag 16.4% (no change)
src/jshim 53.7% (down 1.3% mainly due to new JNI pass through functions for memory manipulation)
src/librdag 91.0% (down 0.3%, largely due to debug statements in `mem.cc`)
src/librdag/runners 95.7% (no change)

### CODE REVIEW

File: src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+   * @param count the number of bytes to set to value
+   * @param value the value to which the memory shall be set
+   */
+  public void memset(long offset, int count, byte value) {
+    checkAccessIsValid(offset, count, 1);
+    Stdlib.memset(_blockStartAddress, offset, count, value);
+  }
+
+  /**
+   * Write an int at a given offset
```
Comment (gmarkall) [inline src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java]: Further to offline discussion: could you change this to "a given byte offset" to minimise ambiguity please?

File: src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+   * @param val the value to write
+   * @param offsetIntoBlock the offset into the block at which to write the value
+   */
+  public void writeInt(int val, long offsetIntoBlock) {
+    checkAccessIsValid(offsetIntoBlock, 1, Stdlib.SIZEOF_JINT);
+    Stdlib.writeIntToAddress(val, _blockStartAddress + offsetIntoBlock);
+  }
+
+  /**
+   * Write a long at a given offset
```
Comment (gmarkall) [inline src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java]: "a given byte offset"

File: src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+   * @param val the value to write
+   * @param offsetIntoBlock the offset into the block at which to write the value
+   */
+  public void writeLong(long val, long offsetIntoBlock) {
+    checkAccessIsValid(offsetIntoBlock, 1, Stdlib.SIZEOF_JLONG);
+    Stdlib.writeLongToAddress(val, _blockStartAddress + offsetIntoBlock);
+  }
+
+  /**
+   * Write a float at a given offset
```
Comment (gmarkall) [inline src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java]: "a given byte offset"

File: src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+   * @param val the value to write
+   * @param offsetIntoBlock the offset into the block at which to write the value
+   */
+  public void writeFloat(float val, long offsetIntoBlock) {
+    checkAccessIsValid(offsetIntoBlock, 1, Stdlib.SIZEOF_JFLOAT);
+    Stdlib.writeFloatToAddress(val, _blockStartAddress + offsetIntoBlock);
+  }
+
+  /**
+   * Write a double at a given offset
```
Comment (gmarkall) [inline src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java]: "a given byte offset"

File: src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+   * @param val the value to write
+   * @param offsetIntoBlock the offset into the block at which to write the value
+   */
+  public void writeDouble(double val, long offsetIntoBlock) {
+    checkAccessIsValid(offsetIntoBlock, 1, Stdlib.SIZEOF_JDOUBLE);
+    Stdlib.writeDoubleToAddress(val, _blockStartAddress + offsetIntoBlock);
+  }
+
+  /**
+   * Read an int from a given offset
```
Comment (gmarkall) [inline src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java]: "a given byte offset"

File: src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+   * @param offsetIntoBlock the offset into the block at which to read the value
+   * @return the read int
+   */
+  public int readInt(long offsetIntoBlock) {
+    checkAccessIsValid(offsetIntoBlock, 1, Stdlib.SIZEOF_JINT);
+    return Stdlib.readIntFromAddress(_blockStartAddress + offsetIntoBlock);
+  }
+
+  /**
+   * Read a long from a given offset
```
Comment (gmarkall) [inline src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java]: "a given byte offset"

File: src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+   * @param offsetIntoBlock the offset into the block at which to read the value
+   * @return the read long
+   */
+  public long readLong(long offsetIntoBlock) {
+    checkAccessIsValid(offsetIntoBlock, 1, Stdlib.SIZEOF_JLONG);
+    return Stdlib.readLongFromAddress(_blockStartAddress + offsetIntoBlock);
+  }
+
+  /**
+   * Read a float from a given offset
```
Comment (gmarkall) [inline src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java]: "a given byte offset"

File: src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+   * @param offsetIntoBlock the offset into the block at which to read the value
+   * @return the read float
+   */
+  public float readFloat(long offsetIntoBlock) {
+    checkAccessIsValid(offsetIntoBlock, 1, Stdlib.SIZEOF_JFLOAT);
+    return Stdlib.readFloatFromAddress(_blockStartAddress + offsetIntoBlock);
+  }
+
+  /**
+   * Read a double from a given offset
```
Comment (gmarkall) [inline src/java/src/main/java/com/opengamma/maths/mem/AVXAlignedMemoryBlock.java]: "a given byte offset"

**Comment by stuartarchibald** [comment]
Comments are now updated in fcfb925 and 7798e21.

**Comment by gmarkall** [comment]
This looks great - thanks for fixing up the comments to make it clear that offsets are in bytes in certain locations.

I see the problem that we need to find a way to get a pointer that won't cause a problem for the arithmetic in Java - however, the search mechanism seems like a workable way around this. Also, given that we have yet to do most of the work for hooking up this interface, and the size of this PR already, I agree that it's a good idea to consider those `exit` calls as a separate case.

Everything looks clean and documented on both the Java and C++ side. The test coverage of the new code looks great - nice work having the patience to be thorough with it.

Happy to merge.

**Pull request merged**